Model

tbohne opened the pull request:
- Single-Player-Mode ("Smart" AI)
- Improved ship placement (Moore-Neighborhood)
- Changed the board representation to a 2D-array (for reasons of clarity)
- Handling of human "dead-end" situations

tbohne committed:
board as 2D-array / check for moore-nh

tbohne committed:
Merge branch 'dev' of github.com:SvantjeJung/battleship into model

tbohne committed:
smart_move() function

tbohne committed:
handling of human dead-end situations

tbohne committed:
merge

tbohne committed:
merged with dev

tbohne committed:
restored client / server placement

tbohne committed:
no need for a dumb ai

tbohne committed:
fixed dead-end bug / removed debugging output

tbohne committed:
further debugging

tbohne committed:
dead-end checks for client and server

tbohne committed:
made misses from opponent visible

SvantjeJung commented on src/client/mod.rs:
```diff
@@ -69,12 +69,16 @@ fn play(mut connection: TcpStream, mut client: Player) {
                                 match res {
                                     MessageType::Hit(id) => {
+                                        let row = id / 10;
+                                        let col = id % 10;
```
There are a lot of lines equivalent to these:
```
let row = X / 10;
let col = X % 10;
```
I think there must be a better solution than always using two new variables.
Must not be changed before presentation, just wanted to take a note on that.

tbohne replied:
I was thinking the same but I have no better solution until now.

SvantjeJung commented on src/main.rs:
```diff
@@ -151,12 +148,11 @@ fn main() {
 
             // TODO: create game instance + AI
```
This line should be deleted now ;)

SvantjeJung commented on src/main.rs:
```diff
@@ -151,12 +148,11 @@ fn main() {
             // TODO: create game instance + AI
             println!("--- Single-Player-Mode ---");
-            mode = Mode::Single;
+            model::start_round();
```
The name is not used anymore. Wouldn't it be nice, if the name of the single-player would be used instead of output like "Player 1, please enter...". So my suggestion would be: `start_round(name)`

SvantjeJung commented on src/model/mod.rs:
```diff
@@ -429,33 +521,60 @@ pub fn place_ships(mut p1: &mut types::Player)
+                        match place(&mut p, i) {
+                            Ok(_) => { break; },
+                            Err(e) => {
+                                match e {
```
High nesting level, but shall suffice for the moment.

SvantjeJung commented on src/model/mod.rs:
```diff
@@ -429,33 +521,60 @@ pub fn place_ships(mut p1: &mut types::Player)
+                                            "please choose another coordinate.");
+                                    },
+                                    types::ErrorType::DeadEndHuman => { return Err(e) },
+                                    _ => { return Err(e) },
```
You could delete the case `types::ErrorType::DeadEndHuman` as the right hand side is equal to the default-case.

SvantjeJung reviewed (commented):
First of all: good work with the AI :)

Not quite sure about the change from `Vec<SubField>` to 2D-SubField-Array. But as the code fixes on a 10x10 board it should be okay for now. Although on direct access the variant `[row][col]` is quite more readable, I think the type `[[SubField; 10]; 10]` is less intuitive as `Vec<SubField>`. One could maybe think about the type alias Board: `type Board = [[SubField; 10]; 10];`. But again: I don't think that must be changed for the presentation...

SvantjeJung committed:
Currentplayer change on every turn

In network game the currentplayer changed only on a Miss.
Single-player-mode changed currentplayer on every turn.
Choosing currentplayer in network game now equals the strategy in
single-player-mode.

SvantjeJung committed:
Some enhancements

+ read board from file for single-player
+ moves of single-player can be made by [a1,1a,A1,1A] instead of only
'A1'
+ name of single-player is used during the game

tbohne committed:
minor changes
